drivers/serial/uart_stm32.h

```diff
@@ -58,6 +58,7 @@ struct uart_stm32_data {
 	uint32_t baud_rate;
 	/* clock device */
 	const struct device *clock;
+	atomic_t tx_lock;
```

Comment:
Shouldn't we use Zephyr spinlock instead of re-implementing it in this driver?

Reply:
Is spinlock available before kernel init ?

Reply:
We don't want to lock interrupt so spinlock isn't adapted.  I used a simple lock because poll_out could be call in interrupt function.

Reply:
> Is spinlock available before kernel init ?

Yes.

Reply:
> We don't want to lock interrupt so spinlock isn't adapted. I used a simple lock because poll_out could be call in interrupt function.

Well, you are checking the nucleo_l432kc which is based on STM32L432xxxx chip which is based on ARM Cortex-M4 which doesn't have atomic instructions (please correct me if I wrong). So, in this case we end up with [atomics-in-C implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/atomic_c.c) which simply uses spinlock internally. And for single core Zephyr spinlock is just interrupt lock :)

Reply:
> > We don't want to lock interrupt so spinlock isn't adapted. I used a simple lock because poll_out could be call in interrupt function.
> 
> Well, you are checking the nucleo_l432kc which is based on STM32L432xxxx chip which is based on ARM Cortex-M4 which doesn't have atomic instructions (please correct me if I wrong). So, in this case we end up with [atomics-in-C implementation](https://github.com/zephyrproject-rtos/zephyr/blob/main/kernel/atomic_c.c) which simply uses spinlock internally. And for single core Zephyr spinlock is just interrupt lock :)

Yes, you're right but I don't understand what you want to do when you say "use Zephyr spinlock instead of re-implementing it in this drive". The function [z_impl_atomic_cas](https://github.com/zephyrproject-rtos/zephyr/blob/f758daa14bb3bfd7dcfd2f6101a0c8444108a56c/kernel/atomic_c.c#L81) protect correctly my lock variable against the uart interrupt  thanks spinlock.

Reply:
@evgeniy-paltsev Do you have more comments ?